src/ucp/wireup/select.c

```diff
@@ -1565,6 +1635,49 @@ ucp_wireup_select_params_init(ucp_wireup_select_params_t *select_params,
     select_params->show_error    = show_error;
 }
 
+/* Lane for doing keepalive */
+static ucs_status_t
+ucp_wireup_add_keepalive_lane(const ucp_wireup_select_params_t *select_params,
+                              ucp_err_handling_mode_t err_mode,
+                              ucp_wireup_select_context_t *select_ctx)
+{
+    ucp_ep_h ep                          = select_params->ep;
+    ucp_worker_h worker                  = ep->worker;
+    ucp_wireup_select_info_t select_info = {0};
+    unsigned ep_init_flags               = ucp_wireup_ep_init_flags(
+                                                   select_params, select_ctx);
+    ucp_wireup_criteria_t criteria;
+    ucs_status_t status;
+
+    if ((err_mode == UCP_ERR_HANDLING_MODE_NONE) ||
+        !ucp_worker_keepalive_is_enabled(worker) || 
+        (select_params->ep_init_flags & UCP_EP_INIT_CREATE_AM_LANE_ONLY)) {
+        return UCS_OK;
+    }
+
+    ucp_wireup_criteria_init(&criteria);
+    criteria.title              = "keepalive";
+    criteria.local_md_flags     = 0;
+    criteria.remote_iface_flags = 0;
+    criteria.local_iface_flags  = 0;
+    criteria.calc_score         = ucp_wireup_am_score_func;
```

Comment:
the keepalive score should also calculate the keepalive scalability

Reply:
1. AM score and KA score are not compared to each other so no need to align them (can reuse common code though)
2. I would do explicit "if" to handle est_num_eps==0 corner case: if est_num_eps>max_inflight_eps, multiply latency by "est_num_eps-max_inflight_eps+1". then the result should be "fed" into am_score as the latency term.